### PR TITLE
[All] Clean lifecycle for v21.06

### PR DESCRIPTION
**Pull request by guparan**
#2197 should be merged first

[ci-depends-on https://github.com/sofa-framework/SofaPython3/pull/150]

Cleaning (removing or setting "deleted") things that have been deprecated for long enough (some of them since 2017...).

Also 
- improved the SOFA_DEPRECATED_HEADER macro to be consistent with SOFA_ATTRIBUTE_DEPRECATED.
- added SOFA_DISABLED_HEADER and SOFA_DISABLED_HEADER_NOT_REPLACED macros
- expanded namespace deprecation in SofaExporter

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

File: SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h [hunk 1]
```diff
@@ -72,7 +72,8 @@ class SOFA_CORE_API ObjectFactory
         /// BaseClass structure associated with the type of intanciated objects.
         virtual const objectmodel::BaseClass* getClass() = 0;
 
-        virtual std::string shortName(objectmodel::BaseObjectDescription* arg) = 0;
+        SOFA_ATTRIBUTE_DISABLED__CLASSNAME_INTROSPECTION()
+        std::string shortName(objectmodel::BaseObjectDescription* arg) = delete;
```
Comment (guparan) [inline SofaKernel/modules/SofaCore/src/sofa/core/ObjectFactory.h:76]: Not 100% sure about the changes in this file :thinking: 
@damienmarchal Could you validate?

**Comment by guparan** [comment]
[ci-build][with-all-tests]

**Comment by epernod** [comment]
241 files +953 −3,830  en fast merge 😉

**Comment by guparan** [comment]
> 241 files +953 −3,830 en fast merge wink

Just to make sure it will be seen because it's needed for v21.06 :smile:

**Comment by guparan** [comment]
[ci-build][with-all-tests]

**Comment by guparan** [comment]
Ready for some "review" (if you are brave enough)! :muscle:

**Review by fredroy — APPROVED** [review]
Looks good to me after super quick glancing over it; and the CI seems happy too so...

**Pull request merged**